### PR TITLE
DAOS-13512 cq: Add a github action for running clang-format.

### PR DESCRIPTION
**Pull request by ashleypittman**
Skip-build: true

Required-githooks: true

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

### CODE REVIEW

**Review by jolivier23 — APPROVED** [review]

File: .github/workflows/clang-format.yml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+# Future work would be to have a way to apply these automatically which would require making and
+# pushing a new git commit so should be optional.
+#
+# Runs in a fedora docker container to get a recent clang-format version.
+
+on:
+  pull_request:
+
+jobs:
+  pylint:
```
Comment (daltonbohning) [inline .github/workflows/clang-format.yml:16]: pylint?

File: .github/workflows/clang-format.yml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    name: Clang Format
+    runs-on: ubuntu-22.04
```
Comment (daltonbohning) [inline .github/workflows/clang-format.yml:18]: Curious. How does this relate to `FROM fedora:38` from the docker file?
The action runs on ubuntu but the actual check on fedora?

**Review by daltonbohning — APPROVED** [review]

File: .github/workflows/clang-format.yml (hunk 1 above)
Comment (ashleypittman) [reply .github/workflows/clang-format.yml:16]: Will change next time.

File: .github/workflows/clang-format.yml (hunk 2 above)
Comment (ashleypittman) [reply .github/workflows/clang-format.yml:18]: It runs in GitHub actions, the options are ubuntu of various versions, mac os or Windows, this check runs on ubuntu but uses a fedora Docker container to do the heavy lifting - the clang-format versions are too old on either ubuntu or el8.

**Review by brianjmurrell — COMMENTED** [review]
Are there really no existing actions out there in the Marketplace (i.e. https://github.com/marketplace/actions/clang-format-lint) or otherwise that do this?

Or do we have some unique requirement?

**Comment by ashleypittman** [comment]
> Are there really no existing actions out there in the Marketplace (i.e. https://github.com/marketplace/actions/clang-format-lint) or otherwise that do this?
> 
> Or do we have some unique requirement?

Good question.  We have SDL requirements about who we're allowed to use actions from and I didn't find one with the correct approvals.

I'd like to extend this to have a method to apply changes as a commit if desired so yes, our future requirements go above and beyond what the current Marketplace actions perform.

**Comment by brianjmurrell** [comment]
> Good question. We have SDL requirements about who we're allowed to use actions from

Do we?  Are we abiding those?

> and I didn't find one with the correct approvals.

Where did you look?  If we are going to be limited to which actions we can use, we are going to end up re-inventing a lot of wheels and taking a long time to do that and generally move very slow.  We simply DO NOT have the resources to be rewriting all of the actions that we are going to need.
 
> I'd like to extend this to have a method to apply changes as a commit if desired so yes,

As the description in the above action makes in it's suggestion:

_You probably want to pair this with a GitHub action (such as EndBug/add-and-commit) to commit the changed files._

> our future requirements go above and beyond what the current Marketplace actions perform.

But do they really?

I generally find actions take on the [philosophy of UNIX tools](https://en.wikipedia.org/wiki/Unix_philosophy), so yes, indeed, rather than finding one action that does it all, you will find several actions you can put together to accomplish the bigger picture goal.

That is exactly what the above Marketplace action is suggesting.

**Comment by ashleypittman** [comment]
> > Good question. We have SDL requirements about who we're allowed to use actions from
> 
> Do we? Are we abiding those?

Well no, but we may have to and I don't want to have to re-write this in that eventuality.

> That is exactly what the above Marketplace action is suggesting.

The one you linked to assumes the target branch is correctly formatted and warns if there are regressions in the PR, we need to check only the commit, not the PR.

I'm a big believer in the one-big-commit approach to enforcing coding styles as the long term cost is lower, but as a project we've decided not to go that route hence we're nudging people in the correct direction.  Hopefully we can start enforcing correct style for modified code shortly and then at least we'll be heading in the right direction.

**Pull request merged**